src/stream/offline_stream.rs

```diff
@@ -0,0 +1,90 @@
+use crate::recognizer::offline_recognizer::{OfflineRecognizer, OfflineRecognizerResult};
+use crate::utils;
+use sherpa_rs_sys::{
+    SherpaOnnxAcceptWaveformOffline, SherpaOnnxDestroyOfflineRecognizerResult,
+    SherpaOnnxDestroyOfflineStream, SherpaOnnxGetOfflineStreamResult, SherpaOnnxOfflineStream,
+};
+
+/// It wraps a pointer from C
+pub struct OfflineStream {
+    pub(crate) pointer: *const SherpaOnnxOfflineStream,
+}
+
+impl Drop for OfflineStream {
+    fn drop(&mut self) {
+        self.delete()
+    }
+}
+
+impl OfflineStream {
+    /// Frees the internal pointer of the stream to avoid memory leak.
+    fn delete(&mut self) {
+        unsafe {
+            SherpaOnnxDestroyOfflineStream(self.pointer);
+        }
+    }
+
+    /// The user is responsible to invoke [Self::drop] to free
+    /// the returned stream to avoid memory leak.
+    pub fn new(recognizer: &OfflineRecognizer) -> OfflineStream {
+        recognizer.new_stream()
+    }
+
+    /// Input audio samples for the offline stream.
+    /// Please only call it once. That is, input all samples at once.
+    ///
+    /// `sample_rate` is the sample rate of the input audio samples. If it is different
+    /// from the value expected by the feature extractor, we will do resampling inside.
+    ///
+    /// `samples` contains the actual audio samples. Each sample is in the range [-1, 1].
+    pub fn accept_waveform(&mut self, sample_rate: i32, samples: &[f32]) {
+        unsafe {
+            SherpaOnnxAcceptWaveformOffline(
+                self.pointer,
+                sample_rate,
+                samples.as_ptr(),
+                samples.len() as i32,
+            );
+        }
+    }
+
+    /// Get the recognition result of the offline stream.
+    pub fn get_result(&self) -> Option<OfflineRecognizerResult> {
```

Comment:
Return result instead of option with bail! Macro to emit errors

Reply:
> Return result instead of option with bail! Macro to emit errors

Is there an err type for the appropriate result at the moment

Reply:
Currently we use eyre::{Result, bail}
We can improve it in future in another PR though, I don't see real benefit of using custom errors in the crates at this stage.

Reply:
> Currently we use eyre::{Result, bail} We can improve it in future in another PR though, I don't see real benefit of using custom errors in the crates at this stage.

In rust practice, if you're a library that returns result, it's a good idea to define your own err and result types. This way, sub-documents can see their own error type and related descriptions. The essence is to facilitate the use of the caller.

Reply:
We'll handle it later. see https://github.com/thewh1teagle/sherpa-rs/issues/46